In [1]:
%jsroot on

UsageError: Line magic function `%jsroot` not found.


在$_{}^{137}Cs$放射源旁放置$LaBr_{3}(Ce)$探测器，利用定标器记录每秒间隔内探测器的计数。通过改变探测器与放射源的距离，得到两组不同计数率下的实验数据。其中一组探测器离放射源较近(high_counts.txt)，另一组探测器离放射源较远(low_counts.txt)。

In [ ]:
ifstream file_high("high_counts.txt");ifstream file_low("low_counts.txt");
TH1F *high_tf1f= new TH1F("high", "hist", 100, 0 ,900.);  //新建一个TH1F类用来储存数据,这组数据用于高斯拟合
TH1F *low_tf1f= new TH1F("low", "hist", 15, 0, 15);

string line;      //读取文件行数
double channel, countnumber,R[2];
while (getline(file_high, line)) 
    {
    stringstream ss(line);
    ss >> channel >> R[0];
    if(R[0]!=0.)
    high_tf1f->Fill(R[0]) ;
}
while (getline(file_low, line)) 
    {
    stringstream ss(line);
    ss >> channel >> R[0];
    low_tf1f->Fill(R[0]) ;
}

画出计数率分布

In [ ]:
TCanvas *c1 = new TCanvas;
c1->Divide(2,1);
c1->cd(1);
high_tf1f->Draw();
c1->cd(2);
low_tf1f->Draw();
c1->Draw();

In [ ]:
c1->cd(1);
high_tf1f->Fit("gaus");
c1->cd(2);
low_tf1f->Fit("gaus");
c1->Draw();

 FCN=17.9479 FROM MIGRAD    STATUS=CONVERGED      58 CALLS          59 TOTAL
                     EDM=3.17785e-09    STRATEGY= 1      ERROR MATRIX ACCURATE 
  EXT PARAMETER                                   STEP         FIRST   
  NO.   NAME      VALUE            ERROR          SIZE      DERIVATIVE 
   1  Constant     1.13818e+03   1.55914e+01   2.67585e-02  -6.06891e-06
   2  Mean         6.66025e+02   2.86280e-01   6.06392e-04   1.73985e-05
   3  Sigma        2.57748e+01   2.08422e-01   4.60557e-06  -2.91319e-02
 FCN=131.666 FROM MIGRAD    STATUS=CONVERGED      56 CALLS          57 TOTAL
                     EDM=3.70095e-07    STRATEGY= 1      ERROR MATRIX ACCURATE 
  EXT PARAMETER                                   STEP         FIRST   
  NO.   NAME      VALUE            ERROR          SIZE      DERIVATIVE 
   1  Constant     1.81521e+03   2.71154e+01   1.13694e-01   3.58372e-05
   2  Mean         3.35620e+00   2.20547e-02   1.23580e-04  -1.81646e-02
   3  Sigma        1.82929e+00   

根据高斯拟合数据的结果，较近的探测器中，$\mu =665.988,\sigma = 25.676，\sqrt{\mu} =25.806 $，
较近的探测器中，$\mu =3.35620,\sigma = 1.82929，\sqrt{\mu} =1.83199$

In [ ]:
TF1 *possi_high = new TF1("possi_high","[0]*TMath::Power(x, [1])*(TMath::Exp(-[1])/TMath::Gamma(x))",0,1000);
TF1 *possi_high_1 = new TF1("possi_high_1","1200*TMath::Power(x, 665)*(TMath::Exp(-665)/TMath::Gamma(x))",1,1000);
TF1 *possi_low = new TF1("possi_low","[0]*TMath::Power(x, [1])*(TMath::Exp(-[1])/TMath::Gamma(x))",0,20);
//TF1 *possi_high = new TF1("possi_high","TMath::Poisson([0],[1])");
//TF1 *possi_low  = new TF1("possi_low","TMath::Poisson([0],[1])");
possi_high->SetParNames ("Constant","Mean");
possi_low->SetParNames ("Constant","Mean");
TCanvas *c2 = new TCanvas;
c2->Divide(2,1);
c2->cd(1);
high_tf1f->Draw();
possi_high_1->Draw();
high_tf1f->Fit("possi_high","R");
c2->cd(2);
low_tf1f->Fit("possi_low","R");
low_tf1f->Draw();
c2->Draw();

 FCN=5.47124e+307 FROM MIGRAD    STATUS=FAILED         86 CALLS          87 TOTAL
                     EDM=-nan    STRATEGY= 1  ERROR MATRIX UNCERTAINTY 100.0 per cent
  EXT PARAMETER                APPROXIMATE        STEP         FIRST   
  NO.   NAME      VALUE            ERROR          SIZE      DERIVATIVE 
   1  Constant     0.00000e+00   1.73205e+00  -0.00000e+00   0.00000e+00
   2  Mean         0.00000e+00   1.00000e+00  -0.00000e+00  2.08428e+305
 FCN=1195.7 FROM MIGRAD    STATUS=CONVERGED     119 CALLS         120 TOTAL
                     EDM=1.64747e-07    STRATEGY= 1      ERROR MATRIX ACCURATE 
  EXT PARAMETER                                   STEP         FIRST   
  NO.   NAME      VALUE            ERROR          SIZE      DERIVATIVE 
   1  Constant     3.50601e+03   5.22545e+01   7.07886e-01   1.14843e-05
   2  Mean         2.89824e+00   5.48594e-02   7.43255e-04   1.22318e-02


Warning in <Fit>: Abnormal termination of minimization.


2.在$_{}^{137}Cs$放射源旁放置$LaBr_{3}(Ce)$探测器，记录每秒间隔内探测器的计数, 并记录相邻脉冲的时间间隔(时间单位$\mu s$) 。

In [ ]:
ifstream counts_per_second("counts_per_second.txt");ifstream times_between_sca_pulses("times_between_sca_pulses.txt");
TH1F *counts_tf1f= new TH1F("counts", "hist", 100, 600., 1000.);  //新建一个TH1F类用来储存数据
while (getline(counts_per_second, line)) 
    {
    stringstream ss(line);
    ss >>  R[0];
    counts_tf1f->Fill(R[0]) ;
}
TCanvas *c3 = new TCanvas;
counts_tf1f->Draw();
c3->Draw();

In [ ]:
TH1F *times_tf1f= new TH1F("times", "hist", 100, 0., 10000.); 
TH1F *times5_tf1f= new TH1F("times5", "hist", 100, 0., 50000.);//新建一个TH1F类用来储存数据
double count5=0.;int n=0;
while (getline(times_between_sca_pulses, line)) 
    {
    n++;
    stringstream ss(line);
    ss >>  R[0];
    times_tf1f->Fill(R[0]) ;
    count5+=R[0];
    if(n==4)
    {
        times5_tf1f->Fill(count5) ;
        n=0;count5=0.;
    }
    
}
TCanvas *c4 = new TCanvas;
c4->Divide(2,1);
c4->cd(1);
times_tf1f->Draw();
c4->cd(2);
times5_tf1f->Draw();
c4->Draw();

In [ ]:
c4->cd(1);
times_tf1f->Fit("expo");  // fit with function of f(x) = exp(p0+p1*x)
c4->Draw();

 FCN=141.593 FROM MIGRAD    STATUS=CONVERGED      52 CALLS          53 TOTAL
                     EDM=5.24134e-11    STRATEGY= 1      ERROR MATRIX ACCURATE 
  EXT PARAMETER                                   STEP         FIRST   
  NO.   NAME      VALUE            ERROR          SIZE      DERIVATIVE 
   1  Constant     8.61688e+00   5.35232e-03   2.23767e-05  -4.95422e-04
   2  Slope       -8.13157e-04   3.03968e-06   1.27081e-08  -3.91774e+00
